articles/logging-with-winston/index.md

```diff
@@ -0,0 +1,575 @@
+transaction.js
+
+```js
+// Require logger.js
+const {usersLogger, transactionLogger} = require('./logger');
+
+...........// Your Transaction' code setups.................
+
+// Transaction logs
+transactionLogger.info('Session connected', { session_id: `${req.id}`}, {user_id: `${res._id}`);
+transactionLogger.info('Transaction Initiated', { transaction_id: `${req.transaction_id}`}, {user_id: `${res.user_id}`);
+transactionLogger.info('Transaction completed',  {user_id: `${res.user_id}`, {`${req.transaction_amout}`, `${req.transaction_code}`});
+transactionLogger.error('Transaction Failed', `{${err}`,session_id: `${req.id}`}, {user_id: `${res._id}`);
+```
+
+This makes it easier to manage your logs because there are seperated for specific issues and user and transaction attributes are provided to help narrow it down to a specific user or transaction.
+
+### Other Logging Middleware
+This guide discusses logging using Winston. However, there are other quality middleware logger packages worth mentioning. They include;
+
+1. Morgan
+
+[Morgan](https://www.npmjs.com/package/morgan) is used for logging HTTP web servers. It combines Node.js with Express. Morgan logs responses and requests from an HTTP server. Its log entries are based on the HTTP request and error loglines.
+
+The Logger function access objects such as `req` and `res`.
+
+- `res` - HTTP response object.
+- `req` - HTTP request object.
+
+The function will return a string log entry to the logline.
+
+Morgan is designed to log errors the way servers like [Apache](https://httpd.apache.org/docs/1.3/logs.html) and [Nginx](https://docs.nginx.com/nginx/admin-guide/monitoring/logging/) carry out to the `access-log` or `error-log`.
+
+Here is an example of a Morgan logger:
+
+```js
+const express = require('express')
+const fs = require('fs')
+const morgan = require('morgan')
+const path = require('path')
+const app = express()
+
+// Create a write stream (in append mode)
+const accessLogStream = fs.createWriteStream(path.join(__dirname, 'access.log'), { flags: 'a' })
+
+// Setup the logger
+app.use(morgan('combined', { stream: accessLogStream }))
+
+app.get('/', (req, res) => {
+    res.send('Hello World!')
+})
+app.listen(3000, () => {
+    console.log("Server Listening on port 3000");
+})
+```
+
+[*Code Source*](https://www.npmjs.com/package/morgan#write-logs-to-a-file)
+
+Run the app and hit http://localhost:3000/ on your browser. Morgan will record the following log on the access.log file.
+
+```js
+::1 - - [12/Nov/2020:09:19:46 +0000] "GET / HTTP/1.1" 304 - "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.183 Safari/537.36"
+```
+
+2. Bunyan
+
+[Bunyan](https://www.npmjs.com/package/bunyan) records log entries in JSON format. The format is automatically generated with essential fields such as timestamp and hostname.
+
+Like the Winston logger, create a logger instance and log your message.
+
+```js
+const Bunyan = require("Bunyan");
```

Comment:
Both occurrences of Bunyan on this line need to be lowercase for this code snippet to work.